blog/content/post/2022-05-04-fobnail_provisioning.md

```diff
@@ -0,0 +1,302 @@
+read other posts related to this project by visiting
+[fobnail](https://blog.3mdeb.com/tags/fobnail/) tag.
+
+# Scope of current phase
+
+This phase is about provisioning Fobnail Token itself. The closing point of that
+process is creating a certificate for Token that can be used later after
+attestation, only if attestation finishes successfully. Before that happens,
+Fobnail Token must verify that Platform Owner can be trusted, and prepare an
+input for that certificate.
+
+## Assumptions and flow of Fobnail provisioning
+
+It is assumed that provisioning is performed in a secure environment, over a
+secure channel, and encryption of data sent between Platform Owner and Fobnail
+Token isn't necessary. After all, Platform Owner is able to issue certificates,
+so there are bigger worries than lack of encryption during provisioning if it
+gets compromised.
+
+Communication with Platform Owner happens through CoAP. Fobnail Token is a
+client, same as during communication with Attester. This is done to keep code
+footprint small enough to fit in nRF52840 dongle. As a consequence, all
+communication is initiated by the Token, so it is responsible for deciding which
+of the three flows (Fobnail provisioning, platform provisioning or attestation)
+it should follow.
+
+First resource that Fobnail Token asks for during its provisioning is a Platform
+Owner certificate chain. It has a [strictly defined format](TBD: link to docs)
+and its root must be provided to the token during compilation. Fobnail validates
+this chain, but because it has neither timekeeping nor (secure) networking
+capabilities, it isn't able to check validity periods or CRLs.
+
+Second request uses POST method and includes CSR (certificate signing request).
+Platform Owner responses to that with a certificate generated from that CSR,
+signed by Platform Owner's private key. This certificate is verified against
+chain received in previous step and saved in non-volatile memory for later use.
+
+![Diagram of Fobnail provisioning flow](/img/Fobnail-flows-fobnail-provisioning.png)
+
+## Fobnail Token's key
+
+There are no good crates for generating RSA keys on Cortex-M, and in our tests
+using universal implementation wasn't able to generate 2048-bit key in 5 hours.
+We also couldn't use ARM TrustZone CryptoCell 310 which is included in nRF52840
+for this purpose because it doesn't have open-sourced libraries. Thus, currently
+Fobnail Token generates Curve25519 key, with help of [Trussed](https://trussed.dev/).
+
+It was originally designed to be used in Diffie–Hellman method to generate
+symmetric encryption key for communication between owners of two separate
+Curve25519 keys. Ed25519 is a signature scheme using Curve25519, which adds the
+ability to sign and verify data, including certificates. Unfortunately,
+Curve25519 can't be used as encryption key pair. This is a problem that will
+have to be resolved to unlock full power of Fobnail Token.
```

Comment:
@krystian-hebel are we tracking this problem somewhere publicly?

Reply:
It is now: https://github.com/fobnail/fobnail/issues/40. There are links in summary section to all issues, should I add another link here?

Reply:
@krystian-hebel It is always better to link wherever we talking about that. Sometime it is hard to connect the dots.

Reply:
Added